content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+
+- **Retryable Exceptions**: It retries requests if they result in any exceptions that are considered retryable. These typically include network-related exceptions like connection timeouts or socket exceptions.
+
+It also has argument constructor. We can use it if default configuration is not suitable for us.
+```java
+public Default(long period, long maxPeriod, int maxAttempts)
+
+// use it to create retryer
+new Retryer.Default(1, 100, 10);
+```
+
+While the default retryer provided by Feign covers many common retry scenarios, there are situations where we might want to define a custom retryer. Here are some motivations for defining a custom retryer:
+
+1. **Fine-Grained Control**: If we need **more control over the default retry behavior**, such as specifying a different maximum number of retry attempts or a custom backoff strategy, a custom retryer allows is to tailor the behavior to our specific requirements.
+
+1. **Retry Logic**: In some cases, we might want to **retry requests only for specific response codes or exceptions**. A custom retryer lets us implement our own logic for determining when a retry should occur.
+
+1. **Logging and Metrics**: If we want to **log or collect metrics related to retry attempts**, implementing a custom retryer provides an opportunity to add this functionality.
+
+1. **Integration with Circuit Breakers**: If we are using circuit breaker patterns in conjunction with Feign, a custom retryer can be integrated with the circuit breaker's state to make more informed decisions about when to retry or when to open the circuit.
+
+1. **Non-Standard Retry Strategies**: For scenarios that do not fit the standard retry strategies provided by the default retryer, such as rate-limited APIs or APIs with specific retry requirements, we can define a custom retryer tailored to our use case.
+
+Here's an example of implementing a custom `Retryer` in OpenFeign:
+```java
+public class CalculatorRetryer implements Retryer {
+    /**
+     * millis to wait between retries
+     */
+    private final long period;
+
+    /**
+     * Maximum number of retries
+     */
+    private final int maxAttempts;
+
+    private int attempt = 1;
+
+    @Override
+    public void continueOrPropagate(RetryableException e) {
+        log.info("Feign retry attempt {} of {} due to {} ", attempt, 
+                                                            maxAttempts, 
+                                                            e.getMessage());
+        if (++attempt > maxAttempts) {
+            throw e;
+        }
+        if (e.status() == 401) {
+            try {
+                Thread.sleep(period);
+            } catch (InterruptedException ex) {
+                throw e;
+            }
+        } else {
+            throw e;
+        }
+    }
+
+    @Override
+    public Retryer clone() {
+        return this;
```

Comment:
Is this method defined in the `Retryer` interface? Shouldn't this return a new instance instead of `this`? What does the methods Javadoc say?

Reply:
We have to return `this` instead of creating a new instance, as it is a stateful retryer... it has to return the `maxAttempts`.

Reply:
Yes, it should return a new instance exactly _because_ it is stateful. If you return `this` and someone calls `clone()` and then modifies the configuration, it will change the configuration for _every_ place where `clone()` has been called. I don't believe that was the intention of the `clone()` method.

Reply:
I have updated the retryer for cloning logic, please review.